src/assignment-5/CollateralizedVault.sol

```diff
@@ -0,0 +1,146 @@
+import {Ownable} from "openzeppelin/contracts/access/Ownable.sol";
+
+interface IERC20WithDecimals is IERC20 {
+    function decimals() external view returns (uint8);
+}
+
+/// @title Collateralized Vault
+/// @author davidbrai
+/// @notice A vault that allows to borrow a token against a collateral
+///     If a debt position is not sufficiently collateralized, e.g when the market price
+///     of the collateral goes down, then the user's collateral may be liquidated by the admin.
+///     In case of a liquidation, both the user's debt and collateral are erased.
+/// @dev The Vault uses Chainlink price feeds to determine the value of the collateral compared to the debt
+contract CollateralizedVault is Ownable {
+
+    /// @notice ERC20 token which can be borrowed from the vault
+    IERC20WithDecimals public immutable underlying;
+
+    /// @notice ERC20 token which can be used as collateral
+    IERC20WithDecimals public immutable collateral;
+
+    /// @notice Chainlink priceFeed of underlying / collateral
+    ///     e.g. if underlying is DAI and collateral is WETH, priceFeed is for DAI/WETH
+    AggregatorV3Interface public immutable priceFeed;
+
+    /// @notice deposited collateral per user in the vault, of `collateral` token
+    mapping(address => uint256) public depositedCollateral;
+
+    /// @notice Mapping of current debt per user, of `underlying` token
+    mapping(address => uint256) public debt;
+
+    error TooMuchDebt();
+    error NotEnoughCollateral();
+    error UserDebtIsSufficientlyCollateralized();
+
+    /// @notice Initalizes a new CollateralizedVault
+    /// @param underlying_ ERC20 token which can be borrowed from the vault
+    /// @param collateral_ ERC20 token which can be used as collateral
+    /// @param priceFeed_ Chainlink priceFeed of underlying / collateral
+    constructor(address underlying_, address collateral_, address priceFeed_) {
+        underlying = IERC20WithDecimals(underlying_);
+        collateral = IERC20WithDecimals(collateral_);
+        priceFeed = AggregatorV3Interface(priceFeed_);
+    }
+
+    /// @notice Deposits additional collateral into the vault
+    /// @param collateralAmount amount of `collateral` token to deposit
+    function deposit(uint256 collateralAmount) public {
+        depositedCollateral[msg.sender] += collateralAmount;
+
+        collateral.transferFrom(msg.sender, address(this), collateralAmount);
+    }
+
+    /// @notice Borrows `underlying` token from the Vault
+    ///     Only allowed to borrow up to the value of the collateral
+    /// @param amount The amount of `underlying` token to borrow
+    function borrow(uint256 amount) public {
+        uint256 maxDebt = getMaxAllowedDebt(depositedCollateral[msg.sender]);
+        if (debt[msg.sender] + amount > maxDebt) {
+            revert NotEnoughCollateral();
+        }
+
+        debt[msg.sender] += amount;
+
```

Comment:
If you do the collateralization check after adding `amount` to `debt[msg.sender]` you can use `getRequiredCollateral` and remove `getMaxAllowedDebt` from the contract. One whole less function :D

Check as well how [MakerDAO does the collateralization check](https://github.com/makerdao/dss/blob/17187f7d47be2f4c71d218785e1155474bbafe8a/src/vat.sol#L163).